oak_idl_gen_services/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+        format!("            }}",),
+    ])
+}
+
+fn generate_service_method(rpc_call: &RPCCall) -> Vec<String> {
+    // Note the asymmetry between input and output types; the input type is a reference to an
+    // object, whose buffer is owned by the caller, while the returned object is created by the
+    // implementation of this method, therefore needs to be wrapped in `oak_idl::Message` in order
+    // to transfer its ownership to the caller.
+    vec![format!(
+        "    fn {}(&self, request: &{}) -> oak_idl::Message<{}>;",
+        method_name(rpc_call),
+        request_type(rpc_call),
+        response_type(rpc_call)
+    )]
+}
+
+/// Returns the `method_id` attribute on the [`RPCCall`]. This is used to identify methods in a
+/// backwards compabile way and without relying on their name or their ordering.
+///
+/// The `method_id` attribute must be declared upfront in each input flatbuffer file.
+fn method_id(rpc_call: &RPCCall) -> anyhow::Result<u32> {
+    rpc_call
+        .attributes()
+        .context("could not find any attributes on method")?
+        .iter()
+        .find(|kv| kv.key() == "method_id")
+        .ok_or_else(|| {
+            anyhow::anyhow!(
+                "method_id attribute not found on method {}",
+                rpc_call.name()
+            )
+        })?
+        .value()
+        .context("could not find attribute value")?
+        .parse::<u32>()
+        .map_err(|err| {
+            anyhow::anyhow!(
+                "could not parse method_id attribute on method {}: {}",
+                rpc_call.name(),
+                err
+            )
+        })
+}
+
+/// The type name of the generated Rust client struct.
+fn client_name(service: &Service) -> String {
+    format!("{}Client", unqualified_name(service.name()))
+}
+
+/// The type name of the generated Rust server struct.
+fn server_name(service: &Service) -> String {
+    format!("{}Server", unqualified_name(service.name()))
+}
+
+/// The type name of the generated Rust service struct.
+fn service_name(service: &Service) -> String {
+    unqualified_name(service.name())
+}
+
+/// The method name of the generated Rust client method.
+fn method_name(rpc_call: &RPCCall) -> String {
+    rpc_call.name().to_case(convert_case::Case::Snake)
+}
+
+/// The type name of the generated Rust request struct.
+fn request_type(rpc_call: &RPCCall) -> String {
+    qualified_name(rpc_call.request().name())
+}
+
+/// The type name of the generated Rust response struct.
+fn response_type(rpc_call: &RPCCall) -> String {
+    qualified_name(rpc_call.response().name())
+}
+
+/// Returns the last part of the name, splitting on `.`.
+fn unqualified_name(qualified_name: &str) -> String {
+    qualified_name
+        .split('.')
+        .last()
+        // If split returned an empty iterator, then the input name did not contain any dots, so we
+        // can return it as-is.
+        .unwrap_or(qualified_name)
+        .to_string()
+}
+
+#[test]
```

Comment:
Having tests mixed inbetween normal functions makes it more difficult to follow for me personally.